[dynamo] make some more graph break messages readable in English [2/N]

williamwen42 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #147912
* #147872
* #147494
* #147429
* __->__ #147385

This is for "for some large number Z, make sure the error messages are readable English." - beginning to audit all `unimplemented` sites and making sure that all messages are at least English-readable. Hints may not necessarily be provided.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

williamwen42 committed:
[dynamo] make some more graph break messages readable in English [2/N]

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] make some more graph break messages readable in English [2/N]"


This is for "for some large number Z, make sure the error messages are readable English." - beginning to audit all `unimplemented` sites and making sure that all messages are at least English-readable. Hints may not necessarily be provided.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] make some more graph break messages readable in English [2/N]"


This is for "for some large number Z, make sure the error messages are readable English." - beginning to audit all `unimplemented` sites and making sure that all messages are at least English-readable. Hints may not necessarily be provided.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] make some more graph break messages readable in English [2/N]"


This is for "for some large number Z, make sure the error messages are readable English." - beginning to audit all `unimplemented` sites and making sure that all messages are at least English-readable. Hints may not necessarily be provided.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] make some more graph break messages readable in English [2/N]"


This is for "for some large number Z, make sure the error messages are readable English." - beginning to audit all `unimplemented` sites and making sure that all messages are at least English-readable. Hints may not necessarily be provided.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] make some more graph break messages readable in English [2/N]"


This is for "for some large number Z, make sure the error messages are readable English." - beginning to audit all `unimplemented` sites and making sure that all messages are at least English-readable. Hints may not necessarily be provided.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jansel commented on torch/_dynamo/codegen.py:
```diff
@@ -293,7 +293,20 @@ def gen_fn():
+                        "If this reconstruction graph break occurs while handling another graph break, then resolve the "
+                        "initial graph break.",
+                        "Report an issue to PyTorch if you need reconstrtuction support. Note that many objects that don't have"
+                        "reconstruction rules are fundamentally unreconstructable "
+                        "(e.g. built-in iterator objects like `map`, `zip`, etc.).",
```
I think these are reconstructable actually, doesn't seem fundamental

zou3519 commented on torch/_dynamo/convert_frame.py:
```diff
@@ -535,7 +535,12 @@ def __call__(
+            unimplemented_v2(
+                gb_type="Attempt to trace generator",
+                context="",
+                explanation="Dynamo cannot trace generators as a top-level function.",
```
"top-level function" a bit confusion. How about "Don't torch.compile a generator function directly (it is OK to torch.compile a function that calls a generator)"

zou3519 commented on torch/_dynamo/convert_frame.py:
```diff
@@ -793,7 +798,12 @@ def log_bytecode(
+                    unimplemented_v2(
+                        gb_type="Excessive RestartAnalysis() calls",
+                        context="",
+                        explanation="Dynamo attempted to restart analysis 100+ times.",
```
Some more information here about why this is bad would be good, like "we're recompiling too much, you are probably getting perf hits, so we decided stop compiling"

zou3519 commented on torch/_dynamo/convert_frame.py:
```diff
@@ -962,7 +972,12 @@ def format_func_info(code: CodeType) -> str:
+                unimplemented_v2(
+                    gb_type="Dynamo cache limit exceeded",
+                    context=f"Limit type: {limit_type}",
+                    explanation=f"Dynamo attempted to recompile too many times, exceeding the {limit_type} cache size limit.",
```
Ditto here, some explanation of why the cache limit exists would help

zou3519 commented on torch/_dynamo/guards.py:
```diff
@@ -1691,7 +1691,12 @@ def NN_MODULE(self, guard: Guard):
         else:
-            exc.unimplemented(f"Guard setup for uninitialized class {type(val)}")
+            exc.unimplemented_v2(
+                gb_type="Attempted to guard on uninitialized nn.Module",
```
What is an uninitialized nn.Module?

williamwen42 replied:
Explained in hint below

zou3519 replied:
```
explanation=f"Attempted to setup NN_MODULE guard on unitialized class {type(val)}.",
hints=[],
```


I don't know what an uninitialized class is either

zou3519 replied:
Oh I found it in the next PR: ""Ensure the `nn.Module` subclass instance has called `super().__init__()`.""

zou3519 commented on torch/_dynamo/output_graph.py:
```diff
@@ -486,7 +486,13 @@ def add_backward_state_hook(self, hook: VariableTracker, prefix="hook"):
+                unimplemented_v2(
+                    gb_type="backward_state does not support export",
+                    context="",
+                    explanation="backward_state (data passed to backward hooks for compiled autograd) "
```
This sounds more like... "compiled autograd doesn't work with torch.export"?

zou3519 commented on torch/_dynamo/output_graph.py:
```diff
@@ -994,7 +1000,12 @@ def compile_subgraph(
+            unimplemented_v2(
+                gb_type="Attempt to compile with unrestorable blocks",
+                context="",
+                explanation="Dynamo cannot compile with active Python blocks (i.e. with/try blocks) that cannot be restored.",
```
I have no clue what this means

zou3519 commented on torch/_dynamo/output_graph.py:
```diff
@@ -2115,7 +2136,12 @@ def get_trace_call_log_str():
             if self.parent is not None:
-                unimplemented("Invoking an nn.Module inside HigherOrderOperator")
+                unimplemented_v2(
+                    gb_type="Invoking an nn.Module inside a HigherOrderOperator",
```
I'm confused because this should be supported

zou3519 replied:
Also for all of the HOP ones, we should use the HOP name directly (e.g. autograd.Function) instead of calling it a HigherOrderOperator, users don't know what this is

williamwen42 replied:
Offline discussion w/ @ydwu4: This should be supported - we can remove this graph break then (or at the very least mark that this graph break is not expected/likely a bug).

ydwu4 replied:
This seems to be creating a call_module node inside the hop subgraph. Probably it's here before the in_nn_module flag because we don't want call_module to be in subgraph.  Can probably delete now.

williamwen42 replied:
inline_inbuilt_nn_modules could still be unset - so this graph break could still occur then?

zou3519 replied:
I don't think there's a good reason for the user to set inline_inbuilt_nn_modules=False

zou3519 commented on torch/_dynamo/side_effects.py:
```diff
@@ -186,8 +186,11 @@ def check_allowed_side_effect(self, item):
-            unimplemented(
-                "HigherOrderOperator: Mutating a variable not in the current scope (SideEffects)"
+            unimplemented_v2(
+                gb_type="HigherOrderOperator: Mutating a variable not in the current scope (SideEffects)",
+                context="",
+                explanation="This is not supported.",
+                hints=[],
```
Can we do something like: 
"Mutating a variable not in the current scope inside of {HOP_NAME} is not supported"

williamwen42 replied:
Will have to figure out how to plumb the HOP name in

zou3519 replied:
Doesn't need to happen in this PR (can you file an issue for a follow-up?)

zou3519 commented on torch/_dynamo/side_effects.py:
```diff
@@ -202,12 +205,22 @@ def load_attr(self, item, name, deleted_ok=False, check=False):
+            unimplemented_v2(
+                gb_type="Write to immutable cell",
+                context=f"cellvar: {cellvar}, value: {value}",
+                explanation="Dynamo doesn't support writing to immutable cell variables.",
```
Why are some cell variables immutable? Are they immutable in cpython or is this a Dynamo concept?

williamwen42 replied:
cc @StrongerXi this looks like a Dynamo concept - we don't support mutating sourceless cells?

zou3519 replied:
If this is a Dynamo concept, then we need to explain these to users differently

StrongerXi replied:
Oh yeah super edge case, I think the "right solution" here is for us Dynamo developers to audit the sources of this, and either fix them or provide a workaround to user. I should've dropped a note to this code block here. Maybe we should just create an issue.

https://github.com/pytorch/pytorch/blob/1c26a95ae3fcb347ebdf36ef7719e6d819d0487f/torch/_dynamo/variables/functions.py#L336-L343

zou3519 commented on torch/_dynamo/side_effects.py:
```diff
@@ -574,7 +592,12 @@ def codegen_save_tempvars(self, cg: PyCodegen):
                 if isinstance(var, variables.AutogradFunctionContextVariable):
-                    unimplemented("AutogradFunctionContextVariable escaped")
+                    unimplemented_v2(
+                        gb_type="AutogradFunctionContextVariable escaped Dynamo-traced region",
```
This sounds more like... "we cannot reconstruct a torch.autograd.Function's ctx object"

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -757,7 +766,12 @@ def wrapper(self: "InstructionTranslatorBase", inst: Instruction):
                     excp.remove_from_stats()
-                    unimplemented("Graph break under GenericContextWrappingVariable")
+                    unimplemented_v2(
+                        gb_type="Graph break under GenericContextWrappingVariable",
+                        context="",
+                        explanation="Attempted to graph break in an active context manager that doesn't support graph breaking.",
```
This is more that... "when there is a graph break in code under a generic context manager, we will graph break on the generic context manager".

Also, it would be nice if the error stack trace points to the graph break instead of the context manager

williamwen42 replied:
Will be addressed in PR 3/N

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -866,7 +880,12 @@ def __init__(cls, name, bases, dct) -> None:
+                gb_type="Missing bytecode handler",
+                context=opname,
+                explanation=f"Dynamo does not know how to handle the bytecode instruction {opname}",
+                hints=[],
```
"file a github issue" ?

williamwen42 replied:
Hint provided in generic graph break hints PR.

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -1315,7 +1346,12 @@ def STORE_GLOBAL(self, inst):
         if isinstance(value, RemovableHandleVariable):
-            unimplemented("Storing handles in globals - NYI")
+            unimplemented_v2(
+                gb_type="Storing handle in globals",
```
what kind of a handle is this?

williamwen42 committed:
Update on "[dynamo] make some more graph break messages readable in English [2/N]"


This is for "for some large number Z, make sure the error messages are readable English." - beginning to audit all `unimplemented` sites and making sure that all messages are at least English-readable. Hints may not necessarily be provided.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/side_effects.py:
```diff
@@ -723,7 +746,12 @@ def codegen_update_mutated(self, cg: PyCodegen):
+                        gb_type="Side effect on existing deque with limited maxlen",
+                        context="",
+                        explanation="This is not supported.",
+                        hints=[],
```
"don't use a deque"? lol

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -1879,7 +1968,12 @@ def CALL_FUNCTION_EX(self, inst):
                 # usecase is uncommon, graph break.
-                unimplemented("random_ op is called with from keyword")
+                unimplemented_v2(
+                    gb_type="random_ op called with `from` keyword",
```
What is a "random_ op"? `Tensor.random_` ?

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -1892,7 +1986,12 @@ def CALL_FUNCTION_EX(self, inst):
                 # usecase is uncommon, graph break.
-                unimplemented("uniform_ op is called with from keyword")
+                unimplemented_v2(
+                    gb_type="uniform_ op called with `from` keyword",
```
ditto here, Tensor.uniform_ or something else?

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -2005,7 +2109,13 @@ def STORE_ATTR(self, inst):
+                gb_type="Should not compile partial graph (STORE_ATTR)",
+                context="",
+                explanation="Dynamo has determined when encountering an unsupported "
+                "STORE_ATTR instruction that it should not compile the partial graph.",
+                hints=[],
```
Is there python code that corresponds to a STORE_ATTR?

williamwen42 committed:
Update on "[dynamo] make some more graph break messages readable in English [2/N]"


This is for "for some large number Z, make sure the error messages are readable English." - beginning to audit all `unimplemented` sites and making sure that all messages are at least English-readable. Hints may not necessarily be provided.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/side_effects.py:
```diff
@@ -218,7 +231,12 @@ def load_cell(self, cellvar):
         if cellvar.pre_existing_contents:
             return cellvar.pre_existing_contents
-        unimplemented("cannot read uninitialized cell")
+        unimplemented_v2(
+            gb_type="Read uninitialized cell",
+            context=str(cellvar),
+            explanation="Cannot read uninitialized cell.",
+            hints=[],
+        )
```
What's an uninitialized cell?

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -540,7 +534,13 @@ def generic_jump(truth_fn: typing.Callable[[object], bool], push: bool):
+            unimplemented_v2(
+                gb_type="Should not compile partial graph (data-dependent jump)",
+                context="",
+                explanation="Dynamo has determined when encountering a data-dependent "
```
How does this manifest from the user side? Do you have example python code that would demonstrate this? I personally do not know what a "data-dependent jump" is so I'm thinking users won't either.

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -1219,9 +1238,21 @@ def LOAD_FAST(self, inst):
+                        context=name,
+                        explanation=f"Could not find an implicit local variable with name {name}",
+                        hints=[
+                            "This happens in dict/list comprehensions",
```
Do you have an example? "LOAD_FAST" still isn't human readable

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -1584,22 +1641,37 @@ def RAISE_VARARGS(self, inst):
+        unimplemented_v2(
+            gb_type="RERAISE in Python < 3.11",
+            context="",
+            explanation="RERAISE bytecode not supported in Python < 3.11",
```
what is a reraise bytecode? How does the user encounter this?

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -2066,7 +2181,12 @@ def BUILD_LIST_UNPACK(self, inst, cls=ListVariable):
+                unimplemented_v2(
+                    gb_type="Failed to unpack object for BUILD_LIST_UNPACK",
+                    context=str(seq),
+                    explanation=f"{seq} cannot be unpacked into a list for the BUILD_LIST_UNPACK bytecode.",
```
What is BUILD_LIST_UNPACK? Is it `*lst`?

zou3519 reviewed (commented):
Mostly looks good to me. Happy to stamp to unblock. My main feedback is that "{bytecode} not supported" error messages isn't understandable for users (who don't know bytecode) - having some python code that may trigger the bytecode in the hints would help, if that's at all possible

williamwen42 committed:
Update on "[dynamo] make some more graph break messages readable in English [2/N]"


This is for "for some large number Z, make sure the error messages are readable English." - beginning to audit all `unimplemented` sites and making sure that all messages are at least English-readable. Hints may not necessarily be provided.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 commented:
@pytorchbot merge

williamwen42 committed:
Update on "[dynamo] make some more graph break messages readable in English [2/N]"


This is for "for some large number Z, make sure the error messages are readable English." - beginning to audit all `unimplemented` sites and making sure that all messages are at least English-readable. Hints may not necessarily be provided.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] make some more graph break messages readable in English [2/N]"


This is for "for some large number Z, make sure the error messages are readable English." - beginning to audit all `unimplemented` sites and making sure that all messages are at least English-readable. Hints may not necessarily be provided.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]
